server/debugsession.go

```diff
@@ -0,0 +1,836 @@
+					Format: "unknown source: {path}",
+					Variables: map[string]string{
+						"path": path,
+					},
+				},
+			))
+		} else {
+			ds.send(&dap.SourceResponse{
+				Response: newDAPSuccessResponse(request.GetRequest()),
+				Body: dap.SourceResponseBody{
+					Content: code,
+				},
+			})
+		}
+
+	case *dap.ContinueRequest:
+		ds.stop = nil
+		ds.debugger.Continue()
+
+		ds.send(&dap.ContinueResponse{
+			Response: newDAPSuccessResponse(request.GetRequest()),
+			Body: dap.ContinueResponseBody{
+				AllThreadsContinued: true,
+			},
+		})
+
+		// NOTE: ContinuedEvent not expected:
+		//   Please note: a debug adapter is not expected to send this event in response
+		//   to a request that implies that execution continues, e.g. ‘launch’ or ‘continue’.
+		//   It is only necessary to send a ‘continued’ event if there was no previous request that implied this.
+
+	case *dap.EvaluateRequest:
+		// TODO: reply with error if ds.stop == nil
+
+		variableName := request.Arguments.Expression
+
+		activation := ds.debugger.CurrentActivation(ds.stop.Interpreter)
+
+		variable := activation.Find(variableName)
+		if variable != nil {
+			value := variable.GetValue()
+			ds.send(&dap.EvaluateResponse{
+				Response: newDAPSuccessResponse(request.GetRequest()),
+				Body: dap.EvaluateResponseBody{
+					Result: value.String(),
+				},
+			})
+		}
+
+		ds.send(newDAPErrorResponse(
+			request.Seq,
+			request.Command,
+			dap.ErrorMessage{
+				Format: "unknown variable: {name}",
+				Variables: map[string]string{
+					"name": variableName,
+				},
+			},
+		))
+
+	case *dap.ScopesRequest:
+		// TODO: return more fine-grained scopes
+
+		ds.send(&dap.ScopesResponse{
+			Response: newDAPSuccessResponse(request.GetRequest()),
+			Body: dap.ScopesResponseBody{
+				Scopes: []dap.Scope{
+					{
+						Name:               "Variables",
+						PresentationHint:   "locals",
+						VariablesReference: int(ScopeIdentifierLocal),
+					},
+					{
+						Name:               "Storage",
+						PresentationHint:   "registers",
+						VariablesReference: int(ScopeIdentifierStorage),
+					},
+				},
+			},
+		})
+
+	case *dap.VariablesRequest:
+		// TODO: reply with error if ds.stop == nil
+		if ds.stop == nil {
+			ds.send(newDAPErrorResponse(request.GetRequest().GetSeq(), "", dap.ErrorMessage{}))
+
+		}
+
+		variableRequested := request.Arguments.VariablesReference
+		responseVariables := make([]dap.Variable, 0)
+
+		inter := ds.stop.Interpreter
+
+		switch ScopeIdentifier(variableRequested) {
+
+		case ScopeIdentifierLocal:
+			//reset variables
+			ds.variableHandleCounter = 0
+			ds.variables = make(map[int]any, 0)
+
+			activation := ds.debugger.CurrentActivation(inter)
+			location := ds.stop.Interpreter.Location
+			functionValues := activation.FunctionValues()
+
+			for name, variable := range functionValues {
+
+				if location.String() == ds.scriptID && name == "self" {
+					continue
+				}
+				if name == "BLS" || name == "RLP" {
+					continue
+				}
```

Comment:
It would be good to generalize that a bit more, so it stays in sync with Cadence

Reply:
Yeah this one (especially self in script context) I spent too long time digging. 

Maybe we can add few interfaces to Value for making debugging part easier.

Reply:
Sounds good. Adding a TODO